### PR TITLE
UCT: tag matching API

### PR DESCRIPTION
**Pull request by alex-mikheev**
@yosefe please take a look

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3378/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/911/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2872/ for details (Mellanox internal link).

File: src/ucp/tag/tag_recv.c [hunk 1]
```diff
@@ -52,7 +52,7 @@ ucp_tag_search_unexp(ucp_worker_h worker, void *buffer, size_t count,
                                                buffer, count, datatype,
                                                &req->recv.state, info);
                 ucs_trace_req("release receive descriptor %p", rdesc);
-                uct_iface_release_am_desc(rdesc);
+                uct_iface_release_desc(rdesc);
```
Comment (yosefe) [inline src/ucp/tag/tag_recv.c]: this function rename should be in a separate PR

File: src/uct/api/tl.h [hunk 2]
```diff
@@ -157,6 +168,30 @@ typedef struct uct_iface_ops {
...
+    ucs_status_ptr_t (*ep_tag_rndv_zcopy)(uct_ep_h ep, uct_tag_t tag,
+                                          const void *header,
+                                          unsigned header_length,
+                                          const uct_iov_t *iov,
+                                          size_t iovcnt,
+                                          uct_completion_t *comp);
+
+    ucs_status_t (*ep_tag_rndv_cancel)(uct_ep_h ep, void *op);
+
+    ucs_status_t (*ep_tag_rndv_request)(uct_ep_h ep, uct_tag_t tag,
```
Comment (yosefe) [inline src/uct/api/tl.h]: i think we don't need this any more because it's done with eager?

File: src/uct/api/tl.h [hunk 3]
```diff
@@ -157,6 +168,30 @@ typedef struct uct_iface_ops {
 
     ucs_status_t (*ep_fence)(uct_ep_h ep, unsigned flags);
 
+    /* Tagged operations */
+
+    ucs_status_t (*ep_tag_eager_short)(uct_ep_h ep, uct_tag_t tag,
```
Comment (yosefe) [inline src/uct/api/tl.h]: can we also add the inline wrapper functions in uct.h for tag operations?

File: src/uct/api/uct_def.h [hunk 4]
```diff
@@ -235,4 +238,59 @@ typedef size_t (*uct_pack_callback_t)(void *dest, void *arg);
...
+ * @warning If the user became the owner of the @a desc (by returning
+ *          @ref UCS_INPROGRESS) the descriptor must be released later by
+ *          @ref uct_iface_release_desc by the user.
+ *
+ * @retval UCS_OK         - descriptor was consumed, and can be released
+ *                          by the caller.
+ * @retval UCS_INPROGRESS - descriptor is owned by the callee, and would be
+ *                          released later.
+ */
+typedef ucs_status_t (*uct_tag_unexp_rndv_cb_t)(void *arg, void *desc,
```
Comment (yosefe) [inline src/uct/api/uct_def.h]: shall we add also the fields in iface_params?

File: src/uct/api/uct.h [hunk 5]
```diff
@@ -1659,4 +1749,263 @@ UCT_INLINE_API ucs_status_t uct_ep_fence(uct_ep_h ep, unsigned flags)
...
+ *                        received despite the cancel request, or
+ *                        UCS_ERR_CANCELED which means the tag was successfully
+ *                        canceled before it was matched.
+ * @param [in]  force     Whether to report completions to ctx->completed_cb. If
+ *                        nonzero, the cancel is assumed to be successful, and
+ *                        the callback is not called.
+ *
+ * @return UCS_OK  -      The tag is canceled in the transport.
+ */
+UCT_INLINE_API ucs_status_t uct_iface_tag_recv_cancel(uct_iface_h iface,
```
Comment (yosefe) [inline src/uct/api/uct.h]: no need, send operations could be canceled on the protocol level (unlike receives)

File: src/uct/api/tl.h (hunk 2 above)
Comment (yosefe) [inline src/uct/api/tl.h]: but we can keep it to hide implementation details in UCT

File: src/uct/api/uct.h (hunk 5 above)
Comment (shamisp) [inline src/uct/api/uct.h]: where the pending queue resides ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/918/ for details.

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2887/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
Before we start - there is a conflict

File: src/uct/api/uct_def.h [hunk 6]
```diff
@@ -235,4 +238,59 @@ typedef size_t (*uct_pack_callback_t)(void *dest, void *arg);
 typedef void (*uct_unpack_callback_t)(void *arg, const void *data, size_t length);
 
 
+/**
+ * Callback to handle unexpected eager tag.
```
Comment (shamisp) [inline src/uct/api/uct_def.h]: 1. @brief is missing. 
2. please describe constrains and requirements associated with the callback. Does it have to be thread or signal safe ? Can the callback call other communication routines ?

File: src/uct/api/uct_def.h [hunk 7]
```diff
@@ -235,4 +238,59 @@ typedef size_t (*uct_pack_callback_t)(void *dest, void *arg);
...
+ * @retval UCS_INPROGRESS - descriptor is owned by the callee, and would be
+ *                          released later.
+ */
+typedef ucs_status_t (*uct_tag_unexp_eager_cb_t)(void *arg, void *data,
+                                                 size_t length, void *desc,
+                                                 uct_tag_t stag,  uint64_t imm);
+
+
+/**
+ * Callback to handle unexpected rendezvous tag.
```
Comment (shamisp) [inline src/uct/api/uct_def.h]: The same as the above comment

File: src/uct/api/uct.h [hunk 8]
```diff
@@ -1659,4 +1759,263 @@ UCT_INLINE_API ucs_status_t uct_ep_fence(uct_ep_h ep, unsigned flags)
     return ep->iface->ops.ep_fence(ep, flags);
 }
 
+
+/**
+ * @ingroup UCT_TAG
+ * @brief Send tagged message using short eager protocol.
+ *
+ * Eager protocol means that the whole data is sent immediately to the peer.
```
Comment (shamisp) [inline src/uct/api/uct.h]: Please follow UPC API documentation format. This routine ...

File: src/uct/api/uct.h [hunk 9]
```diff
@@ -1659,4 +1759,263 @@ UCT_INLINE_API ucs_status_t uct_ep_fence(uct_ep_h ep, unsigned flags)
...
+ */
+UCT_INLINE_API ucs_status_t uct_ep_tag_eager_short(uct_ep_h ep, uct_tag_t tag,
+                                                   const void *data, size_t length)
+{
+    return ep->iface->ops.ep_tag_eager_short(ep, tag, data, length);
+}
+
+/**
+ * @ingroup UCT_TAG
+ * @brief Send tagged message using bcopy eager protocol.
```
Comment (shamisp) [inline src/uct/api/uct.h]: see above

File: src/uct/api/uct.h [hunk 10]
```diff
@@ -1659,4 +1759,263 @@ UCT_INLINE_API ucs_status_t uct_ep_fence(uct_ep_h ep, unsigned flags)
...
+                                              void *arg)
+{
+    return ep->iface->ops.ep_tag_eager_bcopy(ep, tag, imm, pack_cb, arg);
+}
+
+
+/**
+ * @ingroup UCT_TAG
+ * @brief Send tagged message using zcopy eager protocol.
+ *
```
Comment (shamisp) [inline src/uct/api/uct.h]: see above

File: src/uct/api/uct.h [hunk 11]
```diff
@@ -1659,4 +1759,263 @@ UCT_INLINE_API ucs_status_t uct_ep_fence(uct_ep_h ep, unsigned flags)
...
+                                                   uct_completion_t *comp)
+{
+    return ep->iface->ops.ep_tag_eager_zcopy(ep, tag, imm, iov, iovcnt, comp);
+}
+
+
+/**
+ * @ingroup UCT_TAG
+ * @brief Send tagged message using rendezvous protocol.
+ *
```
Comment (shamisp) [inline src/uct/api/uct.h]: Please do another pass over documentation. UCP API is a good reference.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/933/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2899/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/935/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/949/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2912/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
@shamisp please take another look

File: src/uct/api/uct.h [hunk 12]
```diff
@@ -1788,9 +1789,10 @@ UCT_INLINE_API ucs_status_t uct_ep_tag_eager_short(uct_ep_h ep, uct_tag_t tag,
 
 /**
  * @ingroup UCT_TAG
- * @brief Send tagged message using bcopy eager protocol.
+ * @brief Bcopy eager tagged-send operation.
  *
- * Eager protocol means that the whole data is sent immediately to the peer.
+ * This routine sends a message using buffered copy eager protocol. Eager
```
Comment (shamisp) [inline src/uct/api/uct.h]: I suggest to elaborate on buffered. From the documentation one may think there is no difference between buffered and short. I would add that the user data is cached in the bouncing buffer and then later the data is sent from the buffer. Also please describe the motivation  - a.k.a the bouncing buffer is allocated given network constrains and ready for immediate utilization by HW.

File: src/uct/api/uct.h [hunk 13]
```diff
@@ -1817,13 +1819,14 @@ UCT_INLINE_API ssize_t uct_ep_tag_eager_bcopy(uct_ep_h ep, uct_tag_t tag,
...
- * The input data (which has to be previously registered) in @a iov array of
- * @ref uct_iov_t structures sent to remote side ("gather output"). Buffers
- * in @a iov are processed in array order. This means that the function complete
- * iov[0] before proceeding to iov[1], and so on.
+ * This routine sends a message using zero copy eager protocol. Eager protocol
+ * means that the whole data is sent immediately to the peer. The input data
+ * (which has to be previously registered) in @a iov array of @ref uct_iov_t
+ * structures sent to remote side ("gather output"). Buffers in @a iov are
+ * processed in array order, so the function complete iov[0] before proceeding
+ * to iov[1], and so on.
```
Comment (shamisp) [inline src/uct/api/uct.h]: Again. Please elaborate what is the difference between this one , and two other eager protocols.

File: src/uct/api/uct.h [hunk 14]
```diff
@@ -1928,10 +1932,8 @@ UCT_INLINE_API ucs_status_t uct_ep_tag_rdnv_cancel(uct_ep_h ep, void *op)
  * @ingroup UCT_TAG
  * @brief Send software rendezvous request.
  *
- * Send a rendezvous request only, which will allow the peer to complete the
- * data transfer in software. If the peer has posted corresponding receive it
- * will be reported as incomplete match with both data address and key equal to
- * NULL. Otherwise it will be reported as unexpected.
+ * This routine sends a rendezvous request only, which will allow the peer to
```
Comment (shamisp) [inline src/uct/api/uct.h]: "This routine sends a rendezvous request only, which will allow the peer" - does not read well.

File: src/uct/api/uct.h [hunk 15]
```diff
@@ -1909,8 +1913,8 @@ UCT_INLINE_API ucs_status_ptr_t uct_ep_tag_rndv_zcopy(uct_ep_h ep, uct_tag_t tag
  * @ingroup UCT_TAG
  * @brief Cancel outstanding rendezvous operation.
  *
- * The transport will disregard the outstanding operation without calling
- * completion callback provided in @ref uct_ep_tag_rndv_zcopy.
+ * This routine makes transport disregard the outstanding operation without
```
Comment (shamisp) [inline src/uct/api/uct.h]: makes -> signals the underlaying transport ?
What happens if the callback was already called ?

File: src/uct/api/uct.h [hunk 16]
```diff
@@ -1954,8 +1956,9 @@ UCT_INLINE_API ucs_status_t uct_ep_tag_rndv_request(uct_ep_h ep, uct_tag_t tag,
 
 /**
  * @ingroup UCT_TAG
- * @brief Post a tag to be matched on a transport interface.
+ * @brief Post a tag to a transport interface.
  *
+ * This routine posts a tag to be matched on a transport interface.
```
Comment (shamisp) [inline src/uct/api/uct.h]: please elaborate ? why zcopy ?

**Comment by shamisp** [comment]
@bbenton please take a look as well

File: src/uct/api/uct.h (hunk 12 above)
Comment (yosefe) [inline src/uct/api/uct.h]: The UCT design section describes the various modes (eager,short,zcopy). tag matching functions use the same semantics. Why do we need to describe it again here?

Comment (shamisp) [inline src/uct/api/uct.h]: Each API has to be self containing. For now it is just copy-n-past of the same description across all functions. This is not acceptable.

Comment (yosefe) [inline src/uct/api/uct.h]: Agree, that is why we need to have one place which describes what is "short" "bcopy" and "zcopy", then each API would describe the operations itself, and not elaborate much (but refer to) the data type specification. what do you think?

Comment (shamisp) [inline src/uct/api/uct.h]: Probably the right wait is to define semantics for each type of the operation. Once we define it in terms of semantics probably we should be able to reuse it.

Comment (yosefe) [inline src/uct/api/uct.h]: by "reuse" you mean copy/paste?

**Comment by yosefe** [comment]
@alex-mikheev pls rename uct_am_recv_desc_t as well

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2970/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1007/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2971/ for details (Mellanox internal link).

**Comment by alex-mikheev** [comment]
@shamisp please take a look

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2979/ for details (Mellanox internal link).

File: src/uct/api/uct.h [hunk 17]
```diff
@@ -1764,12 +1766,13 @@ UCT_INLINE_API ucs_status_t uct_ep_fence(uct_ep_h ep, unsigned flags)
...
  * @brief Short eager tagged-send operation.
  *
- * This routine sends a message using short eager protocol. Eager protocol
- * means that the whole data is sent immediately to the peer. The data is
- * provided as buffer and its length, and must not be larger than the
- * corresponding @a max_short value in @ref uct_iface_attr. The immediate
- * value delivered to the receiver is implicitly equal to 0. If it’s required
- * to pass non-zero imm value, @ref uct_ep_tag_eager_bcopy should be used.
+ * This routine sends a message using @ref uct_short_protocol_desc "short"
+ * eager protocol. Eager protocol means that the whole data is sent immediately
```
Comment (shamisp) [inline src/uct/api/uct.h]: I'm not sure if "immediately" is the right terminology for eager. One may confuse it with "in place". @bbenton - what do you think.

Comment (bbenton) [inline src/uct/api/uct.h]: I agree.  We should be more careful about the use of the terms "immediate", "immediately", "imm".  They can have multiple, conflicting, meanings.  In this context, I think it is okay to use "immediately" if it is further clarified as to the meaning.  For example, how about this:
  Eager protocol means that the data can be immediately sent to the receiver without any preceding communication, or handshake, to setup the transfer.

File: doc/doxygen/design.dox [hunk 18]
```diff
@@ -33,11 +33,17 @@
...
  * in accelerators. In terms of communication APIs, UCT defines interfaces for
  * immediate (short), buffered copy-and-send (bcopy), and zero-copy (zcopy)
- * communication operations. The short operations are optimized for small
- * messages that can be posted and completed in place. The bcopy operations
- * are optimized for medium size messages that are typically sent through a
- * so-called bouncing-buffer. Finally, the zcopy operations expose
- * zero-copy memory-to-memory communication semantics. 
+ * communication operations.
+ *
+ * \b Short: This type of operations is optimized for small messages that can be posted and completed
```
Comment (bbenton) [inline doc/doxygen/design.dox]: In this and the bcopy and zcopy paragraphs please change "This type of operations" to "This type of operation"

**Comment by yosefe** [comment]
@shamisp @bbenton fixed

File: src/uct/api/uct_def.h [hunk 19]
```diff
@@ -235,4 +238,77 @@ typedef size_t (*uct_pack_callback_t)(void *dest, void *arg);
 typedef void (*uct_unpack_callback_t)(void *arg, const void *data, size_t length);
 
 
+/**
+ * @ingroup UCT_TAG
+ * @brief Callback to process unexpected eager tagged message.
+ *
+ * This callback is invoked when tagged message sent by eager protocol is
+ * arrived and there is no corresponding tag posted.
```
Comment (bbenton) [inline src/uct/api/uct_def.h]: Minor grammar cleanup:
  This callback is invoked when a tagged message sent by the eager protocol has
  arrived and no corresponding tag has been posted.

File: src/uct/api/uct_def.h [hunk 20]
```diff
@@ -235,4 +238,77 @@ typedef size_t (*uct_pack_callback_t)(void *dest, void *arg);
...
+                                                 size_t length, void *desc,
+                                                 uct_tag_t stag,  uint64_t imm);
+
+
+/**
+ * @ingroup UCT_TAG
+ * @brief Callback to process unexpected rendezvous tagged message.
+ *
+ * This callback is invoked when rendezvous send notification is arrived
+ * and there is no corresponding tag posted.
```
Comment (bbenton) [inline src/uct/api/uct_def.h]: Minor grammar cleanup:
  This callback is invoked when a rendezvous send notification has
  arrived and no corresponding tag has been posted.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1032/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2995/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@bbenton fixed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1033/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2996/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1036/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1040/ for details.

**Comment by shamisp** [comment]
@bbenton please provide final ok

**Comment by bbenton** [comment]
:+1: from Doc/API

**Comment by shamisp** [comment]
good to go !
